### PR TITLE
NMS-14210: Add documentation for SELinux as a requirement to run OpenNMS

### PR DESCRIPTION
**Pull request by Bonrob2**
NMS-14210: Add documentation for SELinux as a requirement to run OpenNMS - added line about SELinux policy permitting binding to the ICMP service.

### All Contributors

* [ ] Have you read our [Contribution Guidelines](https://github.com/OpenNMS/opennms/blob/develop/CONTRIBUTING.md)?
* [ ] Have you (electronically) signed the [OpenNMS Contributor Agreement](https://cla-assistant.io/OpenNMS/opennms)?

### Contribution Checklist

* Please [make an issue in the OpenNMS issue tracker](https://issues.opennms.org) if there isn't one already.<br />Once there is an issue, please:
  1. update the title of this PR to be in the format: `${JIRA-ISSUE-NUMBER}: subject of pull request`
  2. update the JIRA link at the bottom of this comment to refer to the real issue number
  3. prefix your commit messages with the issue number, if possible
  4. once you've created this PR, please link to it in a comment in the JIRA issue
  Don't worry if this sounds like a lot, we can help you get things set up properly.
* **If this code is likely to affect the UI, did you name your branch with `-smoke` in it to trigger smoke tests?**
* If this is a new or updated feature, is there documentation for the new behavior?
* If this is new code, are there unit and/or integration tests?
* If this PR targets a `foundation-*` branch, does it try to avoid changing files in `$OPENNMS_HOME/etc/`?

### What's Next?

A PR should be assigned at least 2 reviewers.  If you know that someone would be a good person to review your code, feel free to add them.

If you need help making additions or changes to the documentation related to your changes, please let us know.

In any case, if anything is unclear or you want help getting your PR ready for merge, please don't hesitate to say something in the comments here,
or in [the #opennms-development chat channel](https://chat.opennms.com/opennms/channels/opennms-development).

Once reviewer(s) accept the PR and the branch passes continuous integration, the PR is eligible for merge.

At that time, if you have commit access (are an OpenNMS Group employee or a member of the OGP) you are welcome to merge the PR when you're ready.
Otherwise, a reviewer can merge it for you.

Thanks for taking time to contribute!

### External References

* JIRA (Issue Tracker): http://issues.opennms.org/browse/NMS-14210

### CODE REVIEW

File: docs/modules/deployment/pages/core/getting-started.adoc [hunk 1]
```diff
@@ -30,6 +30,10 @@ endif::[]
 * Internet access to download the installation packages.
 * A working DNS server, and a localhost and server name that resolve properly.
 * A system user with administrative permissions (`sudo`) to perform installation.
+* SELinux policy permits binding to the ICMP service
```
Comment (mmahacek) [inline docs/modules/deployment/pages/core/getting-started.adoc:33]:
```suggestion
* SELinux policy permitting binding to the ICMP service
* ```

File: docs/modules/deployment/pages/core/getting-started.adoc [hunk 2]
```diff
@@ -30,6 +30,10 @@ endif::[]
...
++
+Note that failure to do so may generate an ICMP permission issue when starting OpenNMS.
+(See https://opennms.discourse.group/t/h29-wont-start-with-permission-error-to-open-icmp-socket/2387/1[H29+ won't start with permission error to open ICMP socket] on Discourse.)
```
Comment (mmahacek) [inline docs/modules/deployment/pages/core/getting-started.adoc:36]: Is the content of that Discourse article something that should be copied into the docs to show users how to identify SELinux conflicts?

Comment (Bonrob2) [reply docs/modules/deployment/pages/core/getting-started.adoc:36]: That's a good idea. We could create a separate topic just under the requirements section to do that. Only issue would be maintaining it in two places, if there are ever any updates to the Discourse article.

Comment (mmahacek) [reply docs/modules/deployment/pages/core/getting-started.adoc:36]: The discourse article could be updated to say "this info has been added to the docs" to direct people here for any other future updates

Comment (Bonrob2) [reply docs/modules/deployment/pages/core/getting-started.adoc:36]: It looks like we should use the "considerations with SELinux" section of the Discourse article and put it here. The rest is more describing the problem that they would see after installation and startup. Could create a troubleshooting section and that info to it.

File: docs/modules/deployment/pages/core/getting-started.adoc [hunk 3]
```diff
@@ -30,6 +30,10 @@ endif::[]
 * Internet access to download the installation packages.
 * A working DNS server, and a localhost and server name that resolve properly.
 * A system user with administrative permissions (`sudo`) to perform installation.
+* A SELinux policy permitting binding to the ICMP service.
```
Comment (emarsh1995) [inline docs/modules/deployment/pages/core/getting-started.adoc:33]:
```suggestion
* A SELinux policy which permits binding to the ICMP service.
```

File: docs/modules/deployment/pages/core/getting-started.adoc [hunk 4]
```diff
@@ -30,6 +30,10 @@ endif::[]
...
++
+Note that failure to do so may generate an ICMP permission issue when starting OpenNMS.
+(See https://opennms.discourse.group/t/h29-wont-start-with-permission-error-to-open-icmp-socket/2387/1[H29+ won't start with permission error to open ICMP socket] on Discourse.)
```
Comment (emarsh1995) [inline docs/modules/deployment/pages/core/getting-started.adoc:36]:
```suggestion
Note that failure to include the appropriate SELinux policy may generate an ICMP permission error when starting OpenNMS (see https://opennms.discourse.group/t/h29-wont-start-with-permission-error-to-open-icmp-socket/2387/1[H29+ won't start with permission error to open ICMP socket] on Discourse).
```

**Review by emarsh1995 — CHANGES_REQUESTED** [review]
Some minor changes!

File: docs/modules/deployment/pages/core/getting-started.adoc (hunk 2 above)
Comment (Bonrob2) [reply docs/modules/deployment/pages/core/getting-started.adoc]: Have added that information. Too much? Too little?

File: docs/modules/deployment/pages/core/getting-started.adoc [hunk 5]
```diff
@@ -39,6 +42,78 @@ endif::[]
 
 include::../time-sync.adoc[]
 
+[[selinux-icmp]]
+=== SELinux and ICMP permissions
+
+As noted in the requirements above, you must have SELinux to run {page-component-title}, as servers that have a SELinux policy configured will deny permission for {page-component-title} to bind to the ICMP service.
```
Comment (mmahacek) [inline docs/modules/deployment/pages/core/getting-started.adoc:48]:
```suggestion
If you have SELinux enabled on your {page-component-title} server, there must be a policy configured to allow {page-component-title} to bind to the ICMP service.
```

File: docs/modules/deployment/pages/core/getting-started.adoc [hunk 6]
```diff
@@ -30,6 +30,9 @@ endif::[]
 * Internet access to download the installation packages.
 * A working DNS server, and a localhost and server name that resolve properly.
 * A system user with administrative permissions (`sudo`) to perform installation.
+* A SELinux policy that permits binding to the ICMP service.
```
Comment (mmahacek) [inline docs/modules/deployment/pages/core/getting-started.adoc:33]: It's possible (though not a good security practice) to disable SELinux.   There also are some operating systems that do not include SELinux and don't require these steps done.
```suggestion
* A SELinux policy that permits binding to the ICMP service.
(If you are using an operating system that includes SELinux)
```

File: docs/modules/deployment/pages/core/getting-started.adoc [hunk 7]
```diff
@@ -39,6 +42,78 @@ endif::[]
 
 include::../time-sync.adoc[]
 
+[[selinux-icmp]]
+=== SELinux and ICMP permissions
+
+As noted in the requirements above, you must have SELinux to run {page-component-title}, as servers that have a SELinux policy configured will deny permission for {page-component-title} to bind to the ICMP service.
+
+To verify that you have SELinux running and enabled, use the `sestatus` command.
+You should see the following:
```
Comment (mmahacek) [inline docs/modules/deployment/pages/core/getting-started.adoc:51]:
```suggestion
If enabled, you should see the following:
```

File: docs/modules/deployment/pages/core/getting-started.adoc [hunk 8]
```diff
@@ -39,6 +42,78 @@ endif::[]
...
+Loaded policy name:             targeted
+Current mode:                   enforcing
+Mode from config file:          enforcing
+Policy MLS status:              enabled
+Policy deny_unknown status:     allowed
+Memory protection checking:     actual (secure)
+Max kernel policy version:      33
+----
+
+Use `audit2allow -a` to verify if {page-component-title} violates SELinux policies
```
Comment (mmahacek) [inline docs/modules/deployment/pages/core/getting-started.adoc:67]:
```suggestion
Use `audit2allow -a` to verify if {page-component-title} violates SELinux policies.
```

File: docs/modules/deployment/pages/core/getting-started.adoc [hunk 9]
```diff
@@ -39,6 +42,78 @@ endif::[]
...
+If the command is missing, run the `dnf install policycoreutils-devel` package.
```
Comment (mmahacek) [inline docs/modules/deployment/pages/core/getting-started.adoc:68]:
```suggestion

NOTE: If the `audit2allow` command is missing, run `dnf install policycoreutils-devel` to install package.
```

**Comment by Bonrob2** [comment]
@indigo423 , would you be able to review this PR? Mark has added you as a reviewer. Thanks!

File: docs/modules/deployment/pages/core/getting-started.adoc [hunk 10]
```diff
@@ -30,6 +30,10 @@ endif::[]
 * Internet access to download the installation packages.
 * A working DNS server, and a localhost and server name that resolve properly.
 * A system user with administrative permissions (`sudo`) to perform installation.
+* A SELinux policy that permits binding to the ICMP service.
+(If you are using an operating system that includes SELinux.)
```
Comment (emarsh1995) [inline docs/modules/deployment/pages/core/getting-started.adoc:34]:
```suggestion
* A SELinux policy that permits binding to the ICMP service (if you are using an operating system that includes SELinux).
```

File: docs/modules/deployment/pages/core/getting-started.adoc [hunk 11]
```diff
@@ -39,6 +43,79 @@ endif::[]
 
 include::../time-sync.adoc[]
 
+[[selinux-icmp]]
+=== SELinux and ICMP permissions
+
+If you have SELinux enabled on your {page-component-title} server, there must be a policy configured to allow {page-component-title} to bind to the ICMP service.
+
+To verify that you have SELinux running and enabled, use the `sestatus` command.
```
Comment (emarsh1995) [inline docs/modules/deployment/pages/core/getting-started.adoc:51]:
```suggestion
To verify that you have SELinux running and enabled, use the `sestatus` console command.
```

File: docs/modules/deployment/pages/core/getting-started.adoc [hunk 12]
```diff
@@ -39,6 +43,79 @@ endif::[]
...
+[source, console]
+----
+#============= unconfined_service_t ==============
+allow unconfined_service_t node_t:icmp_socket node_bind;
+
+============= unconfined_service_t ==============
+allow unconfined_service_t port_t:icmp_socket name_bind;
+----
+
+Create a type enforcement file like this: `JniPing.te`.
```
Comment (emarsh1995) [inline docs/modules/deployment/pages/core/getting-started.adoc:83]:
```suggestion
Create a type enforcement file (for example, `JniPing.te`):
```

File: docs/modules/deployment/pages/core/getting-started.adoc [hunk 13]
```diff
@@ -39,6 +43,79 @@ endif::[]
...
+
+Install the SELinux package and enforce it:
+
+[source, console]
+----
+semodule -i JniPing.pp
+getenforce
+----
+
+You can verify the package with `semodule --list-modules | grep JniPing`.
```
Comment (emarsh1995) [inline docs/modules/deployment/pages/core/getting-started.adoc:117]:
```suggestion
You can verify the package with the following console command: `semodule --list-modules | grep JniPing`.
```

**Review by emarsh1995 — CHANGES_REQUESTED** [review]

File: docs/modules/deployment/pages/core/getting-started.adoc (hunk 11 above)
Comment (Bonrob2) [reply docs/modules/deployment/pages/core/getting-started.adoc:51]: I'd rather not specify all the types of commands, as we would have to update all the docs, and that's not a priority. Also, if you hover over the code sample in the docs, you see the type of command.

**Review by emarsh1995 — APPROVED** [review]
LGTM! Thank you! :D

File: docs/modules/deployment/pages/core/centos-rhel7/selinux.adoc [hunk 14]
```diff
@@ -0,0 +1,3 @@
+
+Documentation missing?
+Click *Edit this Page* in the top-right corner to contribute.
```
Comment (emarsh1995) [inline docs/modules/deployment/pages/core/centos-rhel7/selinux.adoc:3]:
```suggestion
Click *Edit This Page* in the top-right corner to contribute.
```

File: docs/modules/deployment/pages/core/centos-rhel8/selinux.adoc [hunk 15]
```diff
@@ -0,0 +1,70 @@
+
+If you have SELinux enabled on your {page-component-title} server, there must be a policy configured to allow {page-component-title} to bind to the ICMP service.
+
+To verify that you have SELinux running and enabled, use the `sestatus` command.
```
Comment (emarsh1995) [inline docs/modules/deployment/pages/core/centos-rhel8/selinux.adoc:4]:
```suggestion
To verify that you have SELinux running and enabled, use the `sestatus` console command.
```

File: docs/modules/deployment/pages/core/centos-rhel8/selinux.adoc [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+If enabled, you should see the following:
```
Comment (emarsh1995) [inline docs/modules/deployment/pages/core/centos-rhel8/selinux.adoc:5]:
```suggestion
If it's enabled, you should see the following:
```

File: docs/modules/deployment/pages/core/centos-rhel8/selinux.adoc [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+Loaded policy name:             targeted
+Current mode:                   enforcing
+Mode from config file:          enforcing
+Policy MLS status:              enabled
+Policy deny_unknown status:     allowed
+Memory protection checking:     actual (secure)
+Max kernel policy version:      33
+----
+
+Use `audit2allow -a` to verify if {page-component-title} violates SELinux policies.
```
Comment (emarsh1995) [inline docs/modules/deployment/pages/core/centos-rhel8/selinux.adoc:21]:
```suggestion
Use `audit2allow -a` to check if {page-component-title} violates SELinux policies.
```

File: docs/modules/deployment/pages/core/centos-rhel8/selinux.adoc [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+
+NOTE: If the `audit2allow` command is missing, run `dnf install policycoreutils-devel` to install package.
```
Comment (emarsh1995) [inline docs/modules/deployment/pages/core/centos-rhel8/selinux.adoc:23]:
```suggestion
NOTE: If the `audit2allow` command is missing, run `dnf install policycoreutils-devel` to install the package.
```

File: docs/modules/deployment/pages/core/centos-rhel8/selinux.adoc [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+
+Look for entries preventing bind ICMP sockets to bind:
```
Comment (emarsh1995) [inline docs/modules/deployment/pages/core/centos-rhel8/selinux.adoc:25]: This line doesn't read very well; I'm not 100% sure what it's trying to say.

File: docs/modules/deployment/pages/core/debian-ubuntu/selinux.adoc [hunk 20]
```diff
@@ -0,0 +1,3 @@
+
+Documentation missing?
+Click *Edit this Page* in the top-right corner to contribute.
```
Comment (emarsh1995) [inline docs/modules/deployment/pages/core/debian-ubuntu/selinux.adoc:3]:
```suggestion
Click *Edit This Page* in the top-right corner to contribute.
```

File: docs/modules/deployment/pages/core/docker/selinux.adoc [hunk 21]
```diff
@@ -0,0 +1,3 @@
+
+Documentation missing?
+Click *Edit this Page* in the top-right corner to contribute.
```
Comment (emarsh1995) [inline docs/modules/deployment/pages/core/docker/selinux.adoc:3]:
```suggestion
Click *Edit This Page* in the top-right corner to contribute.
```

**Review by emarsh1995 — CHANGES_REQUESTED** [review]

File: docs/modules/deployment/pages/core/centos-rhel8/selinux.adoc (hunk 19 above)
Comment (Bonrob2) [reply docs/modules/deployment/pages/core/centos-rhel8/selinux.adoc:25]: @indigo423 , could you provide some feedback here?

Comment (indigo423) [reply docs/modules/deployment/pages/core/centos-rhel8/selinux.adoc:25]: What it tries to say, the audit2allow command will give the user the output which SELinux policies prevent binding ICMP sockets. These policies need to be explicitly allowed from the user.

File: docs/modules/deployment/pages/core/centos-rhel8/selinux.adoc [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+
+Install the SELinux package and enforce it:
+
+[source, console]
+----
+semodule -i JniPing.pp
+getenforce
+----
+
+You can verify the package with the following command: `semodule --list-modules | grep JniPing`.
```
Comment (indigo423) [inline docs/modules/deployment/pages/core/centos-rhel8/selinux.adoc:70]: The commands above are based on the `audit2allow` command on a CentOS 8 system.

File: docs/modules/deployment/pages/core/docker/selinux.adoc [hunk 23]
```diff
@@ -0,0 +1,3 @@
+
+Documentation missing?
+Click *Edit This Page* in the top-right corner to contribute.
```
Comment (indigo423) [inline docs/modules/deployment/pages/core/docker/selinux.adoc:3]: Docker and SELinux are a different story. We can skip this here for now. We would need some more investigation in a separate JIRA ticket.

**Review by indigo423 — COMMENTED** [review]
Some thoughts from my end. Thanks for migrating it over.

File: docs/modules/deployment/pages/core/docker/selinux.adoc (hunk 23 above)
Comment (mmahacek) [reply docs/modules/deployment/pages/core/docker/selinux.adoc:3]: Agree that this probably isn't necessary on Docker.

File: docs/modules/deployment/pages/core/centos-rhel7/selinux.adoc [hunk 24]
```diff
@@ -0,0 +1,3 @@
+
+Documentation missing?
```
Comment (Bonrob2) [inline docs/modules/deployment/pages/core/centos-rhel7/selinux.adoc:2]:
```suggestion
Is there documentation missing?
```

File: docs/modules/deployment/pages/core/centos-rhel8/selinux.adoc [hunk 25]
```diff
@@ -0,0 +1,70 @@
...
+Policy deny_unknown status:     allowed
+Memory protection checking:     actual (secure)
+Max kernel policy version:      33
+----
+
+Use `audit2allow -a` to check if {page-component-title} violates SELinux policies.
+
+NOTE: If the `audit2allow` command is missing, run `dnf install policycoreutils-devel` to install the package.
+
+Look for entries preventing bind ICMP sockets to bind:
```
Comment (Bonrob2) [inline docs/modules/deployment/pages/core/centos-rhel8/selinux.adoc:25]:
```suggestion
Look for entries that prevent the binding of ICMP sockets:
```

File: docs/modules/deployment/pages/core/centos-rhel8/selinux.adoc (hunk 22 above)
Comment (Bonrob2) [inline docs/modules/deployment/pages/core/centos-rhel8/selinux.adoc:70]:
```suggestion
You can verify the package with the following command: `semodule --list-modules | grep JniPing`.

Note that this command is based on the `audit2allow` command on a CentOS 8 system.
```

File: docs/modules/deployment/pages/core/debian-ubuntu/selinux.adoc [hunk 26]
```diff
@@ -0,0 +1,3 @@
+
+Documentation missing?
```
Comment (Bonrob2) [inline docs/modules/deployment/pages/core/debian-ubuntu/selinux.adoc:2]:
```suggestion
Is there documentation missing?
```

**Review by mmahacek — COMMENTED** [review]
I don't know enough about configuring SELinux to give specific feedback, but this shouldn't be merged with "is documentation missing" placeholder text.   As Ubuntu doesn't ship with SELinux built in, I'd vote to remove that section for now.  Also, I have an unconfirmed feeling that the RHEL7 and RHEL8 instructions are probably very similar to each other.

**Comment by Bonrob2** [comment]
> I don't know enough about configuring SELinux to give specific feedback, but this shouldn't be merged with "is documentation missing" placeholder text. As Ubuntu doesn't ship with SELinux built in, I'd vote to remove that section for now. Also, I have an unconfirmed feeling that the RHEL7 and RHEL8 instructions are probably very similar to each other.

Fair enough. I created the tabs based on Ronny's early comments, with the suggestion that if someone in the community could write up instructions if they so wished. I'm happy to remove it all, but that wouldn't address @indigo423's original feedback. I can just remove everything and add text that says this applies only to CentOS 8.

Screen Shot 2023-01-06 at 11 36 41 AM

**Comment by Bonrob2** [comment]
@mmahacek and @indigo423, I have updated to address the version issue.

**Review by mmahacek — APPROVED** [review]
I like just mentioning it's targeted to Cent/RHEL 8.  If the user has SELinux deployed, they should know how to adapt the commands if necessary.

**Comment by Bonrob2** [comment]
> I like just mentioning it's targeted to Cent/RHEL 8. If the user has SELinux deployed, they should know how to adapt the commands if necessary.

It certainly minimizes having to update things.

**Review by emarsh1995 — APPROVED** [review]
LGTM. Thank you!

**Pull request merged**